File: php/src/mainTest.php
<?php

use PHPUnit\Framework\TestCase;

class MainTest extends TestCase
{
    /**
     * @test
     * @dataProvider inputOutputProvider
     */
    public function testProcessInput($inputFile, $outputFile)
    {
        $inputData = file_get_contents($inputFile);
        $expectedOutput = file_get_contents($outputFile);

        // PHPの内部で標準入力をファイルから読み込むストリームを作成
        $stream = fopen('php://temp', 'r+');
        fwrite($stream, $inputData);
        rewind($stream);

        // 標準入力を一時的に置き換える
        $originalStdin = fopen('php://stdin', 'r');
        $GLOBALS['STDIN'] = $stream;

        // 出力をキャプチャする
        ob_start();
        processInput();
        $actualOutput = ob_get_clean();

        // 標準入力を元に戻す
        fclose($GLOBALS['STDIN']);
        $GLOBALS['STDIN'] = $originalStdin;
        // アサーションの実行
        $this->assertSame($expectedOutput, $actualOutput);
    }

    public function inputOutputProvider()
    {
        $inputDir = '/usr/src/app/common/input/A001/';
        $outputDir = '/usr/src/app/common/output/A001/';

        $testCases = [];
        $inputFiles = glob($inputDir . '*.in');

        foreach ($inputFiles as $inputFile) {
            $baseName = basename($inputFile, '.in');
            $outputFile = $outputDir . $baseName . '.out';
            if (file_exists($outputFile)) {
                $testCases[] = [$inputFile, $outputFile];
            }
        }

        return $testCases;
    }
}

File: rust/src/main_test.rs
use super::process_input;
use std::fs::{self, File};
use std::io::{BufReader, Read};
use std::path::Path;

fn read_to_string<P: AsRef<Path>>(path: P) -> String {
    let mut file = File::open(path).expect("Failed to open file");
    let mut content = String::new();
    file.read_to_string(&mut content)
        .expect("Failed to read file");
    content
}

#[test]
fn test_process_input() {
    let input_dir = "/usr/src/app/common/input/A001";
    let output_dir = "/usr/src/app/common/output/A001";
    for entry in fs::read_dir(input_dir).expect("Failed to read input directory") {
        let entry = entry.expect("Failed to get entry");
        let path = entry.path();
        if path.is_file() {
            let file_name = path.file_name().unwrap().to_str().unwrap();
            let input_file = File::open(&path).expect("Failed to open input file");
            let output_file_name = file_name.replace(".in", ".out");
            let output_file_path = Path::new(output_dir).join(output_file_name);
            let expected_output = read_to_string(&output_file_path);

            let reader = BufReader::new(input_file);
            let mut output = Vec::new();
            process_input(reader, &mut output);

            assert_eq!(
                String::from_utf8(output).unwrap(),
                expected_output,
                "Test failed for file: {}",
                file_name
            );
        }
    }
}
